File: astar.py
import heapq

labirinto = []
dist = []
parent = []

def manhattan(a, b):
    return abs(b[0] - a[0]) + abs(b[1] - a[1])

def a_star(ini, fim, n, m):
    pq = []

    dist[ini[0]][ini[1]] = 0
    heapq.heappush(pq, (manhattan(ini, fim), ini))

    moves = [(1,0), (-1,0), (0,1), (0,-1)]

    while pq:
        d, cur = heapq.heappop(pq)

        if cur == fim:
            return

        if dist[cur[0]][cur[1]] == 10**18:
            continue

        for dx, dy in moves:
            nx = cur[0] + dx
            ny = cur[1] + dy

            if nx < 0 or ny < 0 or nx >= n or ny >= m:
                continue
            if labirinto[nx][ny] == 1:
                continue

            if dist[nx][ny] > dist[cur[0]][cur[1]] + 1:
                dist[nx][ny] = dist[cur[0]][cur[1]] + 1
                parent[nx][ny] = cur
                h = manhattan((nx, ny), fim)
                heapq.heappush(pq, (dist[nx][ny] + h, (nx, ny)))

def solve():
    beginx, beginy = map(int, input().split())
    n, m = map(int, input().split())

    global labirinto, dist, parent

    labirinto = [list(map(int, input().split())) for _ in range(n)]
    dist = [[10**18] * m for _ in range(n)]
    parent = [[(-1, -1) for _ in range(m)] for _ in range(n)]

    a_star((beginx, beginy), (n - 1, m - 1), n, m)

    if dist[n - 1][m - 1] == 10**18:
        print("No path")
        return

    print("D:", dist[n - 1][m - 1])

    path = []
    cur = (n - 1, m - 1)

    while cur != (-1, -1):
        path.append(cur)
        cur = parent[cur[0]][cur[1]]

    path.reverse()

    for x, y in path:
        print(f"({x},{y})", end=" ")
    print()

def main():
    t = int(input())
    while t:
        solve()
        t -= 1

main()

File: astar.cpp
#include <bits/stdc++.h>
using namespace std;


#define f first
#define s second
#define pb push_back
#define nd &&
#define ou ||
#define ll long long
#define pii pair<int,int>
#define pb push_back
#define nl '\n'

vector<vector<int>> labirinto;
vector<vector<int>>dist;
vector<vector<pii>> parent;

int manhattan(pii a, pii b) {
    return abs(b.f - a.f) + abs(b.s - a.s);
}

void astar(pii ini, pii fim, int n, int m) {
    priority_queue<pair<int,pii>, vector<pair<int,pii>>, greater<pair<int,pii>>> pq;
    dist[ini.f][ini.s] = 0;
    pq.push({manhattan(ini, fim), ini});

    vector<pii> moves = {{1,0},{-1,0},{0,1},{0,-1}};

    while(!pq.empty()) {
        auto [d, cur] = pq.top(); pq.pop();
        if(cur == fim) return; 
        if(dist[cur.f][cur.s] == INT_MAX) continue;

        for(auto [dx,dy] : moves) {
            int nx = cur.f + dx, ny = cur.s + dy;
            if(nx<0 || ny<0 || nx>=n || ny>=m) continue;
            if(labirinto[nx][ny] == 1) continue; 

            if(dist[nx][ny] > dist[cur.f][cur.s] + 1) {
                dist[nx][ny] = dist[cur.f][cur.s] + 1;
                parent[nx][ny] = cur;
                int h = manhattan({nx,ny}, fim);
                pq.push({dist[nx][ny] + h, {nx,ny}});

            }
        }
    }
}

void solve() {
    int beginx, beginy; cin >> beginx >> beginy;
    int n, m; cin >> n >> m;

    labirinto.resize(n, vector<int>(m));
    dist.assign(n, vector<int>(m, INT_MAX));
    parent.assign(n, vector<pii>(m, {-1,-1}));

    for(int i = 0; i < n; i++) for(int j = 0; j < m; j++) {
        cin >> labirinto[i][j]; 
    }

    astar({beginx, beginy}, {n-1, m-1}, n, m);

    if(dist[n-1][m-1] == INT_MAX) cout << "No path" << nl;
    else cout << "D: " << dist[n-1][m-1] << nl;

    vector<pii> path;
    pii cur = {n-1, m-1};
    while(cur.f != -1 && cur.s != -1) {
        path.push_back(cur);
        cur = parent[cur.f][cur.s];
    }
    reverse(path.begin(), path.end());


    for(auto [x,y] : path) {
        cout << "(" << x << "," << y << ") ";
    }
    cout << nl;

}

int main() {
    int t; cin >> t;
    while(t--) solve();
}
